Added Snakey Swamp map script.

Soggs opened the pull request:
Creates a ribbon works with a snakey path surrounded by water at the top and bottom.

Soggs committed:
Added Snakey Swamp map script.

Creates a ribbon works with a snakey path surrounded by water at the top and bottom.

Soggs committed:
Minor fixes to comply with automated testing.

SimonFlapse commented on map_gen/maps/snakey_swamp.lua:
```diff
@@ -0,0 +1,97 @@
+
+-- Disable Landfill by default
+local Event = require 'utils.event'
+Event.add(
```
This could have been done with the Event.on_init(handler) this way you could enable landfill in game. Right now it will remove the landfill research everytime someone joins for the first time.

SimonFlapse replied:
Fixed in https://github.com/Refactorio/RedMew/pull/997/commits/4e4945f159845bad1687d2fcf7a7f7f65222c59b

SimonFlapse commented on map_gen/maps/snakey_swamp.lua:
```diff
@@ -0,0 +1,97 @@
+    return t
+end
+
+ore_rectangle = b.apply_effect(ore_rectangle, no_resources)
```
With #994 merged you can now use the b.remove_map_gen_resources(shape) function instead of your no_resources function

SimonFlapse replied:
Fixed in https://github.com/Refactorio/RedMew/pull/997/commits/4e4945f159845bad1687d2fcf7a7f7f65222c59b

SimonFlapse committed:
Merge remote-tracking branch 'upstream/develop' into snakey_swamp

SimonFlapse committed:
Changes to starting resources

SimonFlapse commented:
![image](https://user-images.githubusercontent.com/44922798/66736231-ec4d2900-ee68-11e9-818c-496c76452e71.png)
You would most likely need to increase the resource frequency for this map.

Almost no ores and no oil or uranium within 1024 tiles.

You can do this through the `RS.set_map_gen_settings`
You can look in 'resources.map_gen_settings' for presets.

I've also pushed two commits to this branch, so remember to pull before making any more commits

Soggs committed:
Merge branch 'master' into snakey_swamp

Soggs committed:
Now disables Landfill with Event.on_init()

Soggs committed:
Changed snakey_swamp to use Builders.remove_map_gen_entities_by_filter.

Soggs committed:
Merge branch 'snakey_swamp' of https://github.com/Soggs/RedMew into snakey_swamp

# Conflicts:
#	map_gen/maps/snakey_swamp.lua

Soggs commented:
I know about the resources. Such is the fate of ribbon style worlds but I try to avoid overwriting map gen settings as much as possible. I know you typically do that differently but since I come from the modded angle having to go into the map script and change a lot about resource settings is quite annoying.

Soggs committed:
Snakier_Swamp Initial Commit

Adds another ribbon style map similar to Snakey Swamp with even windier paths.

Soggs commented:
![image](https://user-images.githubusercontent.com/15978306/66746944-8e7a0a80-ee83-11e9-8165-a27229144bc5.png)

To illustrate how Snakier Swamps looks. Number of turns is variable, as is the water border width, path length and width, divider width and tile type.

Soggs committed:
Purging the evil trailing whitespaces.

Soggs commented:
Note for myself: Remove decoratives on the dividers and try shallow water 2 as well as green water.

SimonFlapse commented:
> Note for myself: Remove decoratives on the dividers and try shallow water 2 as well as green water.

I'm trying to introduce a function to the builders that allows you to do this quickly and only for a specific shape. See #998

Soggs committed:
Merge remote-tracking branch 'upstream/develop' into snakey_swamp

Soggs committed:
Added Builders.remove_decoratives(shape) to builders.lua

Builders.remove_decoratives(shape) removes any decoratives in the given shape.
Update the map scripts, snakey_swamp and snakier_swamp to use this to remove decoratives from the dividers.

grilledham commented on map_gen/maps/snakier_swamp.lua:
```diff
@@ -0,0 +1,113 @@
+-- Snakey Swamp by Soggs
```
Shouldn't this be `-- Snakier Swamp by whoever`

grilledham commented on map_gen/maps/snakey_swamp.lua:
```diff
@@ -0,0 +1,89 @@
+local max_width = 50
+local ore_rectangle = b.rectangle(min(path_width - 2, max_height), min(quarter_height - 2, max_width) - 10) -- -10 to allow space for market and other buildings
+local ore_spacing = min(quarter_height, max_height + 2)
+local function amount(a)
+    return function(_, _)
+        return ceil(a / min(path_width - 2, max_height) / min(quarter_height - 2, max_width))
+    end
+end
```
The value returned by the inner function is constant from the inner function's point of view. So this could be 
```lua
local function amount(a)
    local value = ceil(a / min(path_width - 2, max_height) / min(quarter_height - 2, max_width))
    return function()
        return value
    end
end
```
To save having to recalculate the same value at runtime.

Same again for snakier swamp

Soggs replied:
I guess I got too used to compilers doing that for me. :D

grilledham commented:
This looks good to me. Waiting for the merge conflict with #998 to be resolved before I approve.

Soggs committed:
Merge remote-tracking branch 'upstream/develop' into snakey_swamp

Soggs committed:
Deleted Builders.remove_decoratives() from Builders.lua as it was a inferior duplicate of the newly added Builders.remove_map_gen_decoratives().

Swapped Snakey_Swamp and Snakier_Swamp to use Builders.remove_map_gen_decoratives() instead of Builders.remove_decoratives().

Soggs committed:
Changes to Snakier_Swamp as proposed by grilledham.

SimonFlapse commented on map_gen/maps/snakey_swamp.lua:
```diff
@@ -0,0 +1,89 @@
+local ore_rectangle = b.rectangle(min(path_width - 2, max_height), min(quarter_height - 2, max_width) - 10) -- -10 to allow space for market and other buildings
+local ore_spacing = min(quarter_height, max_height + 2)
+local function amount(a)
+    return function(_, _)
```
```suggestion
    return function()
```
Since you're not using any parameters for the return function you can just leave the parameters empty.

SimonFlapse commented on map_gen/maps/snakier_swamp.lua:
```diff
@@ -0,0 +1,114 @@
+local function amount(a)
+    local value = ceil(a /min(path_width-2, max_height) /min(path_width-2, max_width))
+    return
+        function (_, _)
```
```suggestion
        function ()
```
Same thing here.
You may also want to use the same function in both maps, but that is up to you.

SimonFlapse reviewed (approved):
LGTM, you may want to fix my other two comments for consistency.

grilledham commented:
LGTM, plus what Simon said.
